pkg/agent/sysadvisor/plugin/poweraware/power_aware.go

```diff
@@ -0,0 +1,125 @@
+
+	"github.com/kubewharf/katalyst-core/pkg/agent/sysadvisor/metacache"
+	"github.com/kubewharf/katalyst-core/pkg/agent/sysadvisor/plugin"
+	"github.com/kubewharf/katalyst-core/pkg/agent/sysadvisor/plugin/poweraware/capper"
+	capserver "github.com/kubewharf/katalyst-core/pkg/agent/sysadvisor/plugin/poweraware/capper/server"
+	"github.com/kubewharf/katalyst-core/pkg/agent/sysadvisor/plugin/poweraware/controller"
+	"github.com/kubewharf/katalyst-core/pkg/agent/sysadvisor/plugin/poweraware/evictor"
+	evictserver "github.com/kubewharf/katalyst-core/pkg/agent/sysadvisor/plugin/poweraware/evictor/server"
+	"github.com/kubewharf/katalyst-core/pkg/agent/sysadvisor/plugin/poweraware/reader"
+	"github.com/kubewharf/katalyst-core/pkg/config"
+	"github.com/kubewharf/katalyst-core/pkg/metaserver"
+	metricspool "github.com/kubewharf/katalyst-core/pkg/metrics/metrics-pool"
+	"github.com/kubewharf/katalyst-core/pkg/util/general"
+)
+
+const metricName = "poweraware-advisor-plugin"
+
+var PluginDisabledError = errors.New("plugin disabled")
+
+type powerAwarePlugin struct {
+	name       string
+	disabled   bool
+	dryRun     bool
+	controller controller.PowerAwareController
+}
+
+func (p powerAwarePlugin) Name() string {
+	return p.name
+}
+
+func (p powerAwarePlugin) Init() error {
+	if p.disabled {
+		general.Infof("pap is disabled")
+		return PluginDisabledError
+	}
+	general.Infof("pap initialized")
+	return nil
+}
+
+func (p powerAwarePlugin) Run(ctx context.Context) {
+	general.Infof("pap running")
+	p.controller.Run(ctx)
+	general.Infof("pap ran and finished")
+}
+
+func NewPowerAwarePlugin(
+	pluginName string,
+	conf *config.Configuration,
+	_ interface{},
+	emitterPool metricspool.MetricsEmitterPool,
+	metaServer *metaserver.MetaServer,
+	_ metacache.MetaCache,
+) (plugin.SysAdvisorPlugin, error) {
+	emitter := emitterPool.GetDefaultMetricsEmitter().WithTags(metricName)
+
+	// avoid returning error as that would exit whole application
+	// failure of its components should limit affect inside this plugin only
+	// whole service shall be downgraded, but running nonetheless
+	var err error
+	var podEvictor evictor.PodEvictor
+	if conf.Disabled || conf.DisablePowerPressureEvict {
+		podEvictor = evictor.NewNoopPodEvictor()
+	} else {
+		if podEvictor, err = evictserver.NewPowerPressureEvictionPlugin(conf, emitter); err != nil {
+			general.Errorf("pap: failed to create power aware plugin: %v", err)
+		}
+	}
+
+	var powerCapper capper.PowerCapper
+	if conf.Disabled || conf.DisablePowerCapping {
+		powerCapper = capper.NewNoopCapper()
+	} else {
+		if powerCapper, err = capserver.NewPowerCapPlugin(conf, emitter); err != nil {
+			general.Errorf("pap: failed to create power aware plugin: %v", err)
+		}
+	}
+
+	// todo: use the power usage data from malachite
+	// we may temporarily have a local reader on top of ipmi (in dev branch), before malachite is ready
+	// todo: consider plugin fashion for reader to hook in
+	var powerReader reader.PowerReader
```

Comment:
Should we initialize reader to a stub? Otherwise calling any method on it unintentionally would panic.

Reply:
yes a power reader stub is protective; will add it